Filtered_kernel/include/CGAL/Static_filtered_predicate.h

```diff
@@ -127,23 +126,29 @@ class Static_filtered_predicate {
   result_type operator()(const A1& a1, const A2& a2, const A3& a3, const A4& a4) const
   {
     CGAL::Epic_converter<AK> convert;
-    typedef typename Kernel_traits<A1>::type EK;
-    typedef typename Type_mapper<A1,EK,Exact_predicates_inexact_constructions_kernel>::type T1;
+    typedef typename Kernel_traits<A1>::type EK1;
+    typedef typename Type_mapper<A1,EK1,Exact_predicates_inexact_constructions_kernel>::type T1;
     std::pair<T1,bool> aa1 = convert(a1.approx());
```

Comment:
Could we just use `auto` here and skip the complications?

Reply:
The problem is that we only look into this code from time to time and have to understand what happens, and said that I do not find `auto` a good idea.  On the other hand we could try to come  up with a better name for `T1`.

Reply:
can you even use `std::tie`?

Reply:
```c++
// std::pair<Epick::XXX, bool>
auto aa1 = convert(CGAL::approx(a1));
```
seems clear enough to me, but as you say, you are the one who will have to reread this, so your choice.
(I didn't remember this use of `std::pair` in place of `boost::optional`)

Reply:
```c++
using Pair1 = Some_mapper<A1>::type;
Pair1 aa1 = convert(CGAL::approx(a1));
```

Reply:
possible implementation:

```c++
template <typename A>
struct Some_mapper
{
    typedef typename Kernel_traits<A>::type EK;
    typedef typename Type_mapper<A,EK,Exact_predicates_inexact_constructions_kernel>::type T;
    typedef typename T type;
}
```

Reply:
When running the Kernel testsuite I saw that we have a problem with `EK4`  when the fourth argument is a number type as the kernel traits of a number type is the  ` Dummy_kernel`.  That is the same for ` Origin` and its ` EK1` but without consequence.   So we must find the real kernel of one of the argument types ` Ai`.

Reply:
> When running the Kernel testsuite I saw that we have a problem with `EK4` when the fourth argument is a number type as the kernel traits of a number type is the `Dummy_kernel`. That is the same for `Origin` and its `EK1` but without consequence. So we must find the real kernel of one of the argument types `Ai`.

I agree with Marc:
```c++
auto aa1 = convert(CGAL::approx(a1));
```
That has two advantages:
  - use `CGAL::approx` instead of a member function `a1.approx()`,
  - no need to do complex meta-programming to implement what you suggest.

Reply:
I came up with [this ](https://gist.github.com/afabri/9aa601c5a469e362caa1e65f9c851deb) to find a kernel different from ` Dummy_kernel`.    I guess that can be simplified to make ` Kernel_traits`  itself recursive.

Reply:
But I agree with you that we only need the exact kernel type for a return type of something we immediately pass as an argument to another function.   More out of curiosity:  Can one write  `std::pair<auto,bool> a  = convert(,,);`?

Reply:
In more recent C++ you can write `std::pair a = convert(...);` (no explicit template parameters), but I don't think you can partially specify them the way you did. I could have missed some extensions though.

Reply:
> In more recent C++ you can write `std::pair a = convert(...);` (no explicit template parameters), but I don't think you can partially specify them the way you did. I could have missed some extensions though.

No, you are right (`auto` cannot be used that way).

As for using class template argument deduction (`std::pair a = convert(...);`), we cannot use C++17 in CGAL, for now.

Reply:
I could use   `auto a; bool b;`  and `std::tie(a,b)`  but just `auto` is fine as well.

Reply:
> I could use `auto a; bool b;` and `std::tie(a,b)` but just `auto` is fine as well.

No you cannot. `auto a;` will not compile because there is no initialization, and the compiler cannot infer the type of `a`.

In C++17, there would be the [structured binding](https://en.cppreference.com/w/cpp/language/structured_binding) (a powerful replacement of `tie` in the core language of C++17):
```c++
auto [aa1, conversion_ok] = convert(CGAL::approx(a1));
```
but we cannot use C++17, for now.

Reply:
The future is always brighter!!!

Reply:
> `auto [aa1, conversion_ok] = convert(CGAL::approx(a1));`

Indeed Andreas' message reminded me of exactly this. But in C++17, I would expect we stop using `pair<T,bool>` and use `optional<T>` instead as return type of `convert`.